Upgrades to JPackage and jpype.imports

Thrameos opened the pull request:
This PR revised the import system.  It adds a Java support class which can detect what is a Java package and its contents.  This allows for wildcard and recursive imports.  There is some additional overhead at startup to compose the list of available modules.  jpype.imports still provides better diagnostics than JPackage but they now share the same machinery.   This is a fairly significant upgrade.

There is one bug fix.  Apparently JPackage was able to import private and protected classes without any warning.  A few of the tests were in fact supposed to be unreachable.  This bug was corrected.

This is actually one of the oldest of my pull requests as the original date for the package determination code was from JPype 0.6.3.   Each year I have taken it out, dusted it off (or started again after my machine got formatted), tried to get it to work, and then been forced to shelve it because of the issues with working with the pre Java 9 issues.  But finally I came across a technical documentation on the replacements made in Java 9 which contained enough technical specs to make it possible to write it for Java 8.  Thus it is finally complete.

Fixes #537

marscher commented on native/common/include/jp_javaframe.h:
```diff
@@ -348,6 +348,12 @@ class JPJavaFrame
 
 	jint compareTo(jobject obj, jobject obj2);
+
+	jboolean isPackage(string str);
```
Can't we pass a const string reference here?

Thrameos replied:
Fixed

marscher commented on native/java/org/jpype/pkg/JPypePackageManager.java:
```diff
@@ -0,0 +1,322 @@
+    if ("jar".equals(resource.getScheme()))
+    {
+      String[] q = resource.toString().split("!");
+      FileSystem zipfs = getJarFileSystem(URI.create(q[0]));
```
zipfs cannot be null here?

Thrameos replied:
Fixed.

marscher commented on native/java/org/jpype/pkg/JPypePackage.java:
```diff
@@ -0,0 +1,126 @@
+    return out.toArray(new String[out.size()]);
+  }
+
+  static boolean isPublic(Path p)
```
the method returns true if the path contains a public package/class? A short docstring wouldn't hurt. The magic going on here is fascinating :)

Thrameos replied:
Done.

Thrameos replied:
Fascinating because it looks useful, or fascinating because I spent multiple weeks digging through technical documents for over 3 years figuring it out?  😪

marscher commented on native/java/org/jpype/pkg/JPypePackage.java:
```diff
@@ -0,0 +1,126 @@
+      {
+        is.read(buffer3.array());
+        buffer3.rewind();
+        int type = buffer3.get();
```
aint't it a byte?

Thrameos replied:
Yep.  I am too used to files where we get an int with a value of -1, but this throws instead.

marscher commented:
From an engineering point of view. These technical documents are publicly available? What a gigantic effort you have taken.

I admire your endurance on this. It is big step into a bright future. 

On April 11, 2020 10:16:08 PM UTC, Karl Nelson <notifications@github.com> wrote:
>@Thrameos commented on this pull request.>
>>
>>
>>
>> +    return null;>
>+  }>
>+>
>+  public String[] getContents()>
>+  {>
>+    ArrayList<String> out = new ArrayList<>();>
>+    for (String key : contents.keySet())>
>+    {>
>+      Object o = this.getObject(key);>
>+      if (o != null)>
>+        out.add(key);>
>+    }>
>+    return out.toArray(new String[out.size()]);>
>+  }>
>+>
>+  static boolean isPublic(Path p)>
>>
>Fascinating because it looks useful, or fascinating because I spent
>multiple weeks digging through technical documents for over 3 years
>figuring it out?  😪 >
>>
>-- >
>You are receiving this because your review was requested.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/jpype-project/jpype/pull/670#discussion_r407116841

Thrameos commented:
The key document turned out to be

Thrameos commented:
The key document turned out to be JEP 220.  It it was discussing the problems that removing the system and introducing the new jrt system.   I was able to get a pattern that worked for Java 9 before but as we were supporting Java 7 and 8 at the time I didn't pursue it as I wanted something universal.   This document gave me enough keys to open the older version once I was able to open the bootloader resources as well as jar resources.   It is always sad when I discover something was possible much earlier, though the difficulty level is much lower now that we have much more code in Java.  I don't think it actually was possible until Java 7 when the zip file system was introduced.

https://openjdk.java.net/jeps/220
https://docs.oracle.com/javase/7/docs/technotes/guides/io/fsp/zipfilesystemprovider.html

I am slightly worried that I will run out of open files with the current implementation.  I didn't really want to leave the file systems open but to get to the paths requires access to an open filesystem.  

i guess I could close it and then reopen it again when I need it.  But will that invalidate the paths?

marscher commented on project/jpype_java/test/org/jpype/pkg/JPypePackageNGTest.java:
```diff
@@ -0,0 +1,88 @@
+    JPypePackage pkg = new JPypePackage("java.lang", JPypePackageManager.getContentMap("java.lang"));
+    for (Map.Entry<String, Path> e : pkg.contents.entrySet())
+    {
+      JPypePackage.isPublic(e.getValue());
```
is that an assertion?

Thrameos replied:
It is just a check for robustness.  The goal of the statement was to make sure nothing would generate an exception while reading a whole bunch of constant pools.  Unfortunately there is no way to know in jar files if you are actually synced up as the only guaranteed structure is the magic number.

I guess we could toss a bunch of known values in, but it would be better to use a structure we control as java/lang can have anything in it.

Thrameos commented:
I made an attempt at not keeping the Jar files open, but it was really slow (32s per 10k vs 600ms per 10k).  Jar files have to be parse completely when they are opened as file systems due to lack of indexing in the compressed files.  Thus it is pretty prohibitive to reopen them multiple times.  I can add a some better resource management in a second pass after this release.

Thrameos commented:
Fixed the issue with too many jars open.
Fixed the issue with Java 9 mis-compiling Java 8 binaries.

https://github.com/eclipse/jetty.project/issues/3244

marscher commented on native/common/jp_javaframe.cpp:
```diff
@@ -1108,3 +1114,34 @@ jint JPJavaFrame::compareTo(jobject obj, jobject obj2)
+		CallObjectMethodA(m_Context->m_JavaContext.get(), m_Context->m_Context_GetPackageID, &v));
+}
+
+jobject JPJavaFrame::getPackageObject(jobject pkg, string str)
```
const str ref please

marscher commented:
Please merge this one after resolving the conflicts. It is a major improvement! Thanks a lot!
